### PR TITLE
feat: add prometheus metrics collection

### PR DESCRIPTION
**Pull request by deepak1725**
Adding Prometheus Metrics Collection. 

Closes: https://issues.redhat.com/browse/APPAI-1828 

image

### CODE REVIEW

File: bayesian/observability/metrics.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    metrics["requests_duration"].labels(
+        method=method,
+        path_template=path_template,
+        status_code_family=status_code_family,
+    ).observe(duration_s)
+
+
+def get_path_template() -> Optional[str]:
+    """Get the path template for the request, or fall back to plain path."""
+    for route in current_app.url_map.iter_rules():
```
Comment (prashbnair) [inline bayesian/observability/metrics.py:80]: why do you need this?  Is it not the same as returning request.path?

File: bayesian/observability/metrics.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+METRICS_PARAMS = {
+    "requests": (
+        Counter,
+        {
+            "name": "http_requests_total",
+            "documentation": "Total count of requests by method, path, and status code.",
+            "labelnames": ["method", "path_template", "status_code"],
+        },
+    ),
+    "requests_duration": (
```
Comment (prashbnair) [inline bayesian/observability/metrics.py:24]: I think we can get to histograms later, can we just use a gauge to measure time?

File: bayesian/observability/metrics.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    prometheus_multiproc_dir = os.environ.get('PROMETHEUS_MULTIPROC_DIR')
+    registry = CollectorRegistry()
+    if prometheus_multiproc_dir:
+        MultiProcessCollector(registry, path=prometheus_multiproc_dir)
+    return registry
+
+
+def init_metrics(registry: CollectorRegistry) -> Dict[str, MetricWrapperBase]:
+    """Initialize the observability with the registry."""
+    metrics = {}
```
Comment (prashbnair) [inline bayesian/observability/metrics.py:47]: Isnt it easier to just call the constructor on each metric than having a map and iterating through it?

File: bayesian/observability/metrics.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    for name, init_bits in METRICS_PARAMS.items():
+        metric_type, params = init_bits
+        metrics[name] = metric_type(registry=registry, **params)
+    return metrics
+
+
+def emit_response_metrics(status_code: int, metrics: Dict[str, MetricWrapperBase]) -> None:
```
Comment (prashbnair) [inline bayesian/observability/metrics.py:54]: I feel this method will grow as we keep adding metrics. Can this be split for each metric?

**Review by prashbnair — CHANGES_REQUESTED** [review]

File: bayesian/observability/metrics.py (hunk 1 above)
Comment (deepak1725) [reply bayesian/observability/metrics.py:80]: This is basically to differenciate between 20x, 30x vs 404 responses.

File: bayesian/observability/metrics.py (hunk 2 above)
Comment (deepak1725) [reply bayesian/observability/metrics.py:23]: Ack

File: bayesian/observability/metrics.py (hunk 4 above)
Comment (deepak1725) [reply bayesian/observability/metrics.py:52]: Ack.

File: bayesian/observability/metrics.py (hunk 3 above)
Comment (deepak1725) [reply bayesian/observability/metrics.py:45]: i feel this being much easier and clear.. anyway if you have a strong opinion, i can change it to Class based too.

File: bayesian/__init__.py [hunk 5]
```diff
@@ -94,6 +106,27 @@ def access_control_allow_origin(response):
...
+
+
+def log_request_finished(sender, **extra):
+    """Request Finish Signal."""
+    emit_response_metrics(extra.get("response").status_code, METRICS)
+    sender.logger.debug('Request is finished', request.path)
+
+
+request_started.connect(log_request_started, app)
+request_finished.connect(log_request_finished, app)
```
Comment (arajkumar) [inline bayesian/__init__.py:128]: It is not a good approach to track every path, which would result in high cardinality output and waste of storage. Better track only the intended endpoints,  e.g. component-analysis.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: bayesian/__init__.py (hunk 5 above)
Comment (deepak1725) [reply bayesian/__init__.py]: Ack

File: bayesian/observability/metrics.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        },
+    )
+}
+
+
+def get_metrics_registry() -> CollectorRegistry:
+    """Initialize a observability registry."""
+    prometheus_multiproc_dir = os.environ.get('PROMETHEUS_MULTIPROC_DIR')
+    registry = CollectorRegistry()
+    if prometheus_multiproc_dir:
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply bayesian/observability/metrics.py]: Ack

File: Dockerfile [hunk 7]
```diff
@@ -8,10 +8,10 @@ LABEL name="bayesian-api" \
       app-license="Apache 2.0"
 
 ENV LANG=en_US.UTF-8 PYTHONDONTWRITEBYTECODE=1
+ENV PROMETHEUS_MULTIPROC_DIR='/coreapi/bayesian/observability'
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply Dockerfile]: set to /tmp

File: bayesian/observability/metrics.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            "documentation": "Gauge of requests processing time by path (in seconds)",
+            "labelnames": ["method", "path_template", "status_code"],
+        },
+    )
+}
+
+
+def get_metrics_registry() -> CollectorRegistry:
+    """Initialize a observability registry."""
+    prometheus_multiproc_dir = os.environ.get('PROMETHEUS_MULTIPROC_DIR')
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply bayesian/observability/metrics.py]: Ack. Logic changed

File: bayesian/observability/metrics.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            "documentation": "Total count of requests by method, path, and status code.",
+            "labelnames": ["method", "path_template", "status_code"],
+        },
+    ),
+    "requests_duration": (
+        Gauge,
+        {
+            "name": "http_requests_duration",
+            "documentation": "Gauge of requests processing time by path (in seconds)",
+            "labelnames": ["method", "path_template", "status_code"],
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply bayesian/observability/metrics.py]: Ack. Logic changed

File: bayesian/observability/metrics.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import os
+from prometheus_client import CollectorRegistry
+from prometheus_client.multiprocess import MultiProcessCollector
+
+
+METRICS_PARAMS = {
+    "requests": (
+        Counter,
+        {
+            "name": "http_requests_total",
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply bayesian/observability/metrics.py]: Ack. Logic changed

File: bayesian/observability/metrics.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            "documentation": "Total count of requests by method, path, and status code.",
+            "labelnames": ["method", "path_template", "status_code"],
```
[the remark this answers is not in the document]
Comment (deepak1725) [reply bayesian/observability/metrics.py]: Ack. Logic changed

**Comment by deepak1725** [comment]
Closing in favor of https://github.com/fabric8-analytics/fabric8-analytics-server/pull/755

**Pull request closed without merging**